keras_nlp/samplers/beam_sampler.py

```diff
@@ -162,16 +170,23 @@ def gather_beams(x):
         )
 
         # Gather the top beam at each batch index.
-        prompt, log_probs = unflatten_beams(prompt), unflatten_beams(log_probs)
-        top_beams = tf.math.argmax(log_probs, axis=-1)[:, tf.newaxis]
-        prompt = tf.gather(prompt, top_beams, axis=1, batch_dims=1)
-        return tf.squeeze(prompt, axis=1)
+        all_prompts, all_log_probs = unflatten_beams(prompt), unflatten_beams(
+            log_probs
+        )
+        top_beams = tf.math.argmax(all_log_probs, axis=-1)[:, tf.newaxis]
+        prompt = tf.gather(all_prompts, top_beams, axis=1, batch_dims=1)
+
+        if return_all_beams:
```

Comment:
I wonder if instead we should do something like this...

  - By default, return the top beam output sequences with shape `(batch_size, length)`. (as we do currently)
  - If `return_all_beams==True`, return an `(outputs, log_probs)` tuple, with shapes `(batch_size, num_beams, length)` and `(batch_size, num_beams)` respectively, where each beam is ordered so the most likely is first.

This would add a little complexity to the implementation (we would probably need to do an `argsort` and `gather` for the `return_all_beams` branch), but it would make the return type more useful and less redundant to the end user.

Want the top beams? That's `outputs[:, 0, :]`. Want the second to top? That's `outputs[:, 1, :]`.

Reply:
Sounds good! A concern would be how to handle beams which have the same probability. Ensuring that the output order remains the same no matter how many times we generate it over the input sequence is beneficial.

Reply:
I think argmax will break ties by going with whatever came first, so I think we could just use it without adding any randomness to the process.